### PR TITLE
Add HyperOpt Example

### PR DESCRIPTION
**Pull request by Tsmith5151**
- Update requirements.txt with hyperopt library
- Add hyperopt example for XGBoost binary classification to the optimization.ipynb notebook
- Add XGBoostClassifierHyperOpt to optimization.py 
- This class inherits from XGBoostClassifier and uses hyperopt to minimizes the loss function given a stochastic search space (e.g. n_rounds)

### CODE REVIEW

File: slickml/optimization.py [hunk 1]
```diff
@@ -335,3 +339,157 @@ def plot_optimization_results(self):
...
+    xgb_cv: class method
+        Optimization function for XGBoost utilizing cross-validation
+    """
+
+    def __init__(
+        self,
+        params=None,
+        num_boost_rounds=200,
+        metrics="auc",
+        n_split=4,
```
Comment (amirhessam88) [inline slickml/optimization.py:383]: It should be n_splits

File: slickml/optimization.py [hunk 2]
```diff
@@ -335,3 +339,157 @@ def plot_optimization_results(self):
...
+        shuffle=True,
+        early_stop=20,
```
Comment (amirhessam88) [inline slickml/optimization.py:385]: it should be early_stopping_rounds

File: slickml/optimization.py [hunk 3]
```diff
@@ -335,3 +339,157 @@ def plot_optimization_results(self):
...
+            self.early_stop = early_stop
+        else:
+            raise TypeError("The input early_stop must be a int")
+
+        if isinstance(verbose, bool):
+            self.verbose = verbose
+        else:
+            raise TypeError("The input verbose must be a boolean")
+
+    def process(self, X_train, Y_train, func_name, space, trials, algo, max_evals):
```
Comment (amirhessam88) [inline slickml/optimization.py:427]: this should be fit instead of process

File: slickml/optimization.py [hunk 4]
```diff
@@ -335,3 +339,157 @@ def plot_optimization_results(self):
...
+        according to a given algorithm, allowing up to a certain number of
+        function evaluations.
+        Parameters
+        ----------
+        X_train: numpy.array or Pandas DataFrame
+            Training features data
+        y_train: numpy.array[int] or list[int]
+            List of training ground truth binary values [0, 1]
+        func_name: str
+            function name for performing optimization
```
Comment (amirhessam88) [inline slickml/optimization.py:439]: Capital F in Function

File: slickml/optimization.py [hunk 5]
```diff
@@ -335,3 +339,157 @@ def plot_optimization_results(self):
...
+            The set of possible arguments to `fn` is the set of objects
+            that could be created with non-zero probability by drawing randomly
+            from this stochastic program involving involving hp
+        trials: object
+            Storage for completed, ongoing, and scheduled evaluation points.  If
+            None, then a temporary `base.Trials` instance will be created.  If
+            a trials object, then that trials object will be affected by
+            side-effect of this call
+        algo: object
+            provides logic for sequential search of the hyperparameter space
```
Comment (amirhessam88) [inline slickml/optimization.py:450]: Capital P for Provides

File: slickml/optimization.py [hunk 6]
```diff
@@ -335,3 +339,158 @@ def plot_optimization_results(self):
...
+    process: class method
+        Performs hyperparameter tunning on input dataset
+    xgb_cv: class method
+        Optimization function for XGBoost utilizing cross-validation
+    """
+
+    def __init__(
+        self,
+        params=None,
+        num_boost_rounds=200,
```
Comment (amirhessam88) [inline slickml/optimization.py:381]: These all should None and re-define in __init__. Either they should inherit or re-instantiated

File: slickml/optimization.py [hunk 7]
```diff
@@ -335,3 +339,158 @@ def plot_optimization_results(self):
...
+            self.early_stop = early_stop
+        else:
+            raise TypeError("The input early_stop must be a int")
+
+        if isinstance(verbose, bool):
+            self.verbose = verbose
+        else:
+            raise TypeError("The input verbose must be a boolean")
+
+    def process(self, X_train, Y_train, func_name, space, trials, algo, max_evals):
```
Comment (amirhessam88) [inline slickml/optimization.py:427]: The same here. func_name, space, trials, algo, max_evals, these all should have initialized values. 

I checked the example. you have:
def get_xgb_params():
    """ Define Parameter Space"""
    params  = {'nthread':4,
              'booster':'gbtree',
              'tree_method':'hist',
              'objective':'binary:logistic',
              'max_depth':hp.choice('max_depth',range(5,100,5)),
              'learning_rate':hp.quniform('learning_rate',0.01,0.50,0.01),
              'gamma':hp.quniform('gamma',0,0.50,0.01),
              'min_child_weight':hp.quniform('min_child_weight',1,10,1),
              'subsample':hp.quniform('subsample',0.1,1,0.01),
              'colsample_bytree': hp.quniform('colsample_bytree',0.1,1.0,0.01),
              'gamma': hp.quniform('gamma',0.0,1.0,0.01),
              'reg_alpha':hp.quniform('reg_alpha',0.0,1.0,0.01),
              'reg_lambda':hp.quniform('reg_lambda',0.0,1.0,0.01)
              }
    return params

This should be initialized inside the function. with like more logical values. For instance max_depth > 7 like it never happens.

File: slickml/optimization.py [hunk 8]
```diff
@@ -335,3 +339,158 @@ def plot_optimization_results(self):
...
+        self.X_train = X_train
+        self.y_train = Y_train
+        try:
+            results = fmin(
+                fn, space=space, algo=algo, max_evals=max_evals, trials=trials
+            )
+            results = pd.DataFrame(results, index=[0])
+        except Exception as e:
+            return {"status": STATUS_FAIL, "exception": str(e)}
+        return results, trials
```
Comment (amirhessam88) [inline slickml/optimization.py:468]: fit function does not return anything. Therefore helper functions needed to return the outputs.
like get_optimization_results() and get_optimization_trials(). Please do check in Bayesian. So, it will follow the same naming convention.

File: slickml/optimization.py [hunk 9]
```diff
@@ -335,3 +339,158 @@ def plot_optimization_results(self):
...
+            None, then a temporary `base.Trials` instance will be created.  If
+            a trials object, then that trials object will be affected by
+            side-effect of this call
+        """
+
+        fn = getattr(self, func_name)
+        self.X_train = X_train
+        self.y_train = Y_train
+        try:
+            results = fmin(
```
Comment (amirhessam88) [inline slickml/optimization.py:462]: Another thing I noticed that here. FMIN... this is really important. For instance Bayesian Opt always maximize the objective. Therefore, if the metrics are something like AUC, we can just pass it. but if you pass metric as "error" or "logloss", you need to maximize (-1) * metric. The same would be here as well. Please check the documentation. I believe you are minimizing the AUC.

**Review by amirhessam88 — CHANGES_REQUESTED** [review]
